cmd/pj-rehearse/main.go

```diff
@@ -1,23 +1,278 @@
+	rehearsal.Context = fmt.Sprintf("ci/rehearse/%s/%s", repo, strings.TrimPrefix(source.Context, "ci/prow/"))
+
+	if len(source.Spec.Containers) != 1 {
+		return nil, fmt.Errorf("Cannot rehearse jobs with more than 1 container in Spec")
+	}
+	container := source.Spec.Containers[0]
+
+	if len(container.Command) != 1 || container.Command[0] != "ci-operator" {
+		return nil, fmt.Errorf("Cannot rehearse jobs that have Command different from simple 'ci-operator'")
+	}
+
+	for _, arg := range container.Args {
+		if strings.HasPrefix(arg, "--git-ref") || strings.HasPrefix(arg, "-git-ref") {
+			return nil, fmt.Errorf("Cannot rehearse jobs that call ci-operator with '--git-ref' arg")
+		}
+	}
+
+	if len(source.Branches) != 1 {
+		return nil, fmt.Errorf("Cannot rehearse jobs that run over multiple branches")
+	}
+	branch := strings.TrimPrefix(strings.TrimSuffix(source.Branches[0], "$"), "^")
+
+	gitrefArg := fmt.Sprintf("--git-ref=%s@%s", repo, branch)
+	rehearsal.Spec.Containers[0].Args = append(source.Spec.Containers[0].Args, gitrefArg)
+
+	return &rehearsal, nil
+}
+
+func submitRehearsal(job *prowconfig.Presubmit, jobSpec *pjapi.ProwJobSpec, logger logrus.FieldLogger, pjclient pj.ProwJobInterface, dry bool) (*pjapi.ProwJob, error) {
+	labels := make(map[string]string)
+	for k, v := range job.Labels {
+		labels[k] = v
+	}
+
+	pj := pjutil.NewProwJob(pjutil.PresubmitSpec(*job, *(jobSpec.Refs)), labels)
+	logger.WithFields(pjutil.ProwJobFields(&pj)).Info("Submitting a new prowjob.")
+
+	if dry {
+		jobAsYAML, err := yaml.Marshal(pj)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to marshal job to YAML: %v", err)
+		}
+		fmt.Printf("%s\n", jobAsYAML)
+		return &pj, nil
+	}
+
+	return pjclient.Create(&pj)
+}
+
+func execute(toBeRehearsed map[string][]prowconfig.Presubmit, jobSpec *pjapi.ProwJobSpec, logger logrus.FieldLogger, rehearsalConfigs *rehearsalCIOperatorConfigs, pjclient pj.ProwJobInterface, dry bool) error {
+	rehearsals := []*prowconfig.Presubmit{}
+
+	for repo, jobs := range toBeRehearsed {
+		for _, job := range jobs {
+			jobLogger := logger.WithFields(logrus.Fields{"target-repo": repo, "target-job": job.Name})
+			rehearsal, err := makeRehearsalPresubmit(&job, repo, getPrNumber(jobSpec))
+			if err != nil {
+				jobLogger.WithError(err).Warn("Failed to make a rehearsal presubmit")
+			} else {
+				jobLogger.WithField("rehearsal-job", rehearsal.Name).Info("Created a rehearsal job to be submitted")
+				rehearsalConfigs.FixupJob(rehearsal, repo)
+				rehearsals = append(rehearsals, rehearsal)
+			}
+		}
+	}
+
+	if len(rehearsals) > 0 {
+		if err := rehearsalConfigs.Create(); err != nil {
+			return fmt.Errorf("failed to prepare rehearsal ci-operator config ConfigMap: %v", err)
+		}
+		for _, job := range rehearsals {
+			created, err := submitRehearsal(job, jobSpec, logger, pjclient, dry)
+			if err != nil {
+				logger.WithError(err).Warn("Failed to execute a rehearsal presubmit presubmit")
```

Comment:
stutter

Reply:
Addressed in https://github.com/openshift/ci-operator-prowgen/pull/49/commits/d8f71b4ffd2d83717659ce733846c3519d521e86